Consolidated Throttle/Collective into Thrust

ekaszubski opened the pull request:
Resolves #515.

Updated field names in UAVOs:
- ActuatorDesired
- AltitudeHoldState
- LoiterCommand
- StabilizationDesired

Updated references and throttle/collective logic in various flight modules.

This PR also enables helicopters to use flight modes which stabilize altitude.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/d-ronin/dronin/569)

<!-- Reviewable:end -->

ekaszubski committed:
Consolidated Throttle/Collective into Thrust

mlyle commented:
After 496 merges, this should rebase and we'll look at it.

ekaszubski commented:
@mlyle Sounds good. Oh also note that this PR should enable altitude hold, loiter, path following, etc on helicopters.

ekaszubski committed:
Updated license in all files touched by this PR

ekaszubski commented:
Rebasing onto next now that #496 merged :)

mlyle commented:
I'm hoping #473 can sequence first; about to rebase it and cue it up for my testing (no one else hsould play with it yet IMO, I've been pretty sloppy lately)

ekaszubski commented:
OK, sounds good to me. I'm bench-testing this and #581 on my 450 now.

mlyle commented:
I have lightly skimmed this and it looks good to me overall (not promising no bugs and that I won't eventually have small suggested edits)

ekaszubski commented:
@mlyle Much appreciated!

mlyle commented:
Needs rebase, unfortunately.

I've flown this a bunch in ordinary acro and leveling flight on Brain on multirotor.  If it does altitudehold successfully on helicopter I think it's ready for merge.

ekaszubski committed:
Added get_thrust_source function and updated loiter and althold to use it

ekaszubski commented:
Forgot to add thrust logic in loiter and althold functions in transmitter_control during first pass; fixed now.

mlyle commented:
```
/var/lib/jenkins/jobs/dron/workspace@2/Nodes/master/flight/Modules/ManualControl/transmitter_control.c:1181:1: error: expected identifier or '(' before '{' token
 {
 ^
/var/lib/jenkins/jobs/dron/workspace@2/Nodes/master/flight/Modules/ManualControl/transmitter_control.c:1180:13: error: 'altitude_hold_desired' declared 'static' but never defined [-Werror=unused-function]
 static void altitude_hold_desired(ManualControlCommandData * cmd, bool flightModeChanged, SystemSettingsAirframeTypeOptions * airframe_type);
```

ekaszubski commented:
Have a bunch of other stuff going on and kinda rushing through this one, sorry for all the failed builds

mlyle commented:
no worries, doesn't bug me :)

jenkins hates you now though.

ekaszubski commented:
Haha oh well, finally passed build. I'll start testing this on my 480 tomorrow.

ekaszubski committed:
Fixed any negative thrust being forced to full-negative for collective

ekaszubski committed:
Fixed collective input scaling for althold and loiter

ekaszubski committed:
Set disarmed failsafe thrust position for helicp to 0 for neutral collective

ekaszubski committed:
Added helicp-specific failsafe parameters

ekaszubski committed:
Made actuator look at throttle instead of collective for positiveThrottle

ekaszubski committed:
Made lowThrottleZeroIntegral look at throttle instead of collective

ekaszubski commented:
Fixed some heli-specific bugs, multi performance should be unchanged. Changes were mostly
- situations where modules were looking at thrust when they really wanted throttle data
- situations where modules were doing something special when throttle went negative (and were looking at thrust instead of throttle)

Tests:
- [x] Bench test non-armed failsafe to verify collective and throttle outputs
  - success
- [x] Bench test armed failsafe to verify collective and throttle outputs
  - success
- [x] Bench test failsafe flight mode to verify collective and throttle outputs
  - RPY are set correctly but throttle and collective are still able to be controlled; investigating
  - fixed
- [x] Bench test althold command of collective on 450-size heli
  - success
- [x] Flight test althold on multirotor to check for regressions (thanks @droidicus)
  - no issues observed
- [x] Flight test althold command of collective on 450-size heli
  - success
  - this is the first time I've flown anything in altitude hold mode
  - transition in/out of althold was nice and smooth
  - altitude control seemed very accurate; flew less than a meter off the ground in hover and forward flight
  - full climb rate was reasonable, full descent rate felt slow but still controllable

ekaszubski committed:
Throttle and collective are now correctly set in failsafe flight mode

droidicus commented:
Tested next + #620 + #569 on a sparky2 150mm quad with 1104s, in both leveling and altitudehold. Flew well, no issues noted.

ekaszubski commented:
@droidicus Thanks for the test. I'm aiming to head out to the field early tomorrow morning to test this on my 450 and possibly my 480 (with GPS)

ekaszubski commented:
Video of test on 450-size heli: https://youtu.be/-0HAp5Hvrqs

ekaszubski commented on flight/Modules/ManualControl/geofence_control.c:
```diff
@@ -91,7 +91,7 @@ int32_t geofence_control_select(bool reset_controller)
 		/* Pick default values that will roughly cause a plane to circle down and */
 		/* a quad to fall straight down */
-		stabilization_desired.Throttle = -1;
+		stabilization_desired.Thrust = -1;
```
Looks like this should be updated to behave the same as failsafe_control i.e. set thrust to 0 if helicp, -1 otherwise (to prevent helicp from shooting downward when geofence is active). On the other hand, geofence is pretty much guaranteed to end in a bad crash for helicp with or without this fix.

ekaszubski commented:
@mlyle  Additional fixes are being posted to https://github.com/ekaszubski/dronin/commits/thrust-axis-fixes and can be made into another PR or combined with this one, up to you.

mlyle commented:
Flown on brain, including althold.

Flown on sparky2.
